src/pages/signup.js

```diff
@@ -0,0 +1,260 @@
+      setForm({ ...form, message: { text: "Error signing up with email and password", type: "error" } })
+      return false
+    }
+    return true
+  }
+
+  const submitForm = (event) => {
+    event.preventDefault()
+    if (validateInputs()) {
+      sendUserVerificationEmail(event)
+    }
+  }
+
+  // the provider sign-in
+  useEffect(() => {
+    // The user will either sign-in with email-link or with provider. Check each conditionally.
+
+    // Confirm the link is a sign-in with email link.
+    if (auth.isSignInWithEmailLink(window.location.href)) {
+      setForm({ ...form, onVerifyLink: true })
+
+      // Get the email if available. This should be available if the user completes
+      // the flow on the same device where they started it.
+      let email = window.localStorage.getItem("emailForSignIn")
+      if (!email) {
+        // User opened the link on a different device. To prevent session fixation
+        // attacks, ask the user to provide the associated email again. For example:
+        email = window.prompt("Please provide your email for confirmation")
+      }
+      // The client SDK will parse the code from the link for you.
+      auth
+        .signInWithEmailLink(email, window.location.href)
+        .then((result) => {
+          // Clear email from storage.
+          window.localStorage.removeItem("emailForSignIn")
+
+          prepareUserInformation(result.user).then((res) => {
+            getCurrentUser()
+              .getIdToken(true)
+              .then((idToken) => {
+                dispatch({ type: "LOGIN", user: res, idt: idToken })
+                navigate("/")
+              })
+          })
+        })
+        .catch((error) => {
+          // Some error occurred, you can inspect the code: error.code
+          // Common errors could be invalid email and invalid or expired OTPs.
+          console.log("error signing in with emailLink", error)
+        })
+    } else {
+      auth
+        .getRedirectResult()
+        .then((result) => {
+          if (result.user) {
+            prepareUserInformation(result.user).then((res) => {
+              getCurrentUser()
+                .getIdToken(true)
+                .then((idToken) => {
+                  dispatch({ type: "LOGIN", user: res, idt: idToken })
+                  navigate("/")
+                })
+            })
+          } else {
+            // only get rid of the spinner if a provider AND a verify email link sign-in are not being attempted
+            if (!form.onVerifyLink) {
+              setForm({ ...form, isLoading: false })
+            }
+          }
+        })
+        .catch((error) => {
+          setForm({ ...form, isLoading: false })
+          if (error.code === "auth/account-exists-with-different-credential") {
+            setForm({ ...form, message: { text: "An account already exists under that email address", type: "error" } })
+          }
+        })
+    }
+  }, [])
```

Comment:
Codacy found an issue: [React Hook useEffect has missing dependencies: 'dispatch' and 'form'. Either include them or remove the dependency array. You can also do a functional update 'setForm(f => ...)' if you only need 'form' in the 'setForm' call.](https://app.codacy.com/gh/treetoplearning/heartwood/pullRequest?prid=5919976)